storage/remote/metadata_watcher.go

```diff
@@ -0,0 +1,158 @@
+// Copyright 2020 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package remote
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/scrape"
+)
+
+// MetadataAppender is an interface used by the Metadata Watcher to send metadata read from the scrape manager on to somewhere else.
+type MetadataAppender interface {
+	AppendMetadata(context.Context, []scrape.MetricMetadata)
+}
+
+// Watchable is an interface used to represent a scrape manager from which we fetch metadata.
+type Watchable interface {
+	TargetsActive() map[string][]*scrape.Target
+}
+
+// MetadataWatcher watches the Scrape Manager for a given WriteMetadataTo.
+type MetadataWatcher struct {
+	name   string
+	logger log.Logger
+
+	managerGetter scrape.ManagerGetter
+	manager       Watchable
+	writer        MetadataAppender
+
+	interval model.Duration
+	deadline time.Duration
+
+	quit chan struct{}
+	done chan struct{}
+
+	hardShutdown context.CancelFunc
+}
+
+// NewMetadataWatcher builds a new MetadataWatcher.
+func NewMetadataWatcher(l log.Logger, mg scrape.ManagerGetter, name string, w MetadataAppender, interval model.Duration, deadline time.Duration) *MetadataWatcher {
+	if l == nil {
+		l = log.NewNopLogger()
+	}
+
+	if mg == nil {
+		mg = &scrape.ReadyManager{}
+	}
+
+	return &MetadataWatcher{
+		name:   name,
+		logger: l,
+
+		managerGetter: mg,
+		writer:        w,
+
+		interval: interval,
+		deadline: deadline,
+
+		quit: make(chan struct{}),
+		done: make(chan struct{}),
+	}
+}
+
+// Start the MetadataWatcher.
+func (mw *MetadataWatcher) Start() {
+	level.Info(mw.logger).Log("msg", "Starting scraped metadata watcher")
+	var hardShutdownCtx context.Context
+	hardShutdownCtx, mw.hardShutdown = context.WithCancel(context.Background())
+
+	go mw.loop(hardShutdownCtx)
+}
+
+// Stop the MetadataWatcher.
+func (mw *MetadataWatcher) Stop() {
+	level.Info(mw.logger).Log("msg", "Stoping metadata watcher...")
```

Comment:
I think we stick to lower case mesages log :thinking:

Reply:
This is consistent with what I see in other places e.g. [discovery manager, scrape manager, rule manager](https://github.com/prometheus/prometheus/blob/babcb86cd07ff177b59aed2745f63e353f3f37d7/cmd/prometheus/main.go#L502-L575) where starting and stopping messages are uppercase.

Do you still think it's worth lowercasing?

Reply:
Yea, I am not sure, I think we should be consistent. I know in Thanos we ended up having lowercase as it's machinery (structured) log at the end, but don't remember if we had agreement on this on Prometheus, worth to check on IRC?

Reply:
Will check.

Reply:
Hows the investigation? ;p Not a big deal

Reply:
I have filed https://github.com/prometheus/prometheus/issues/6972 to help clear this. 😄

Reply:
Per the issue above, the casing is correct. If you don't agree, let's discuss that into the issue itself.